src/core/socket.c

```diff
@@ -141,11 +143,20 @@ void socket_free_ports(Socket *s) {
 
 static void socket_done(Unit *u) {
         Socket *s = SOCKET(u);
+        SocketPeer *p;
+        Iterator i;
 
         assert(s);
 
         socket_free_ports(s);
 
+        HASHMAP_FOREACH(p, s->peers_by_address, i) {
+                hashmap_remove(s->peers_by_address, p);
+                free(p);
+        }
```

Comment:
this will explode, no? I mean what happens if a Service object referencing one of these SocketPeer objects is cleaned up after the Socket object? in that case its reference to the SocketPeer will point to an object that is already freed.

I think it would be best to allow SocketPeer to exist either attached to a Socket, or detached from it. The latter would happen only if the Socket it originally belonged to got destroyed, but the SocketPeer is still referenced by a Socket. I'd encode the fact whether a SocketPeer object is attached or detached simply by including a ptr from the SocketPeer to its Socket object, which would be NULL if the object is detached, and non-NULL if still attached.